kernel/src/snapshot.rs

```diff
@@ -546,4 +659,646 @@ mod tests {
             Some(3)
         );
     }
+
+    #[test]
```

Comment:
Aside: @nicklan -- This file is getting pretty large. I wonder what the rust best practice is for where unit tests should live? 

My experience from non-rust projects is that bulky unit tests should generally live in separate (test-only) source files -- especially tests that only use public interfaces. In-file testing would make more sense for exercising private/internal code that isn't accessible outside the source file.